Update serving code to enable `saved_model=True`

amyeroberts opened the pull request:
# What does this PR do?

Fixes and adds any missing`serving` and `serving_output` code to our TF models to enable 
`model.save_pretrained(path, saved_model=True)`

I've added comments throughout the code to explain any areas where the TF refactor might no be obvious. 

I'm aware the diff of this PR is quite big, but most of it repetitive changes to enable passing of tests, so I hope acceptable. 

## Specifically: 
**1. Adds missing serving logic to: ResNet, Swin, TAPAS**

**2. Adds missing `serving_output` logic to models** 
Some vision models didn't have `serving_output` implemented - `serving` returned the model outputs directly. This was to enable testing (see 4.) and to keep consistent with the rest of the library, 

**3. Update or add `input_signature` decorator for models**
**4. Adds a test to check `serving_output` is implemented and return types are as expected**

We can't test `model.serving` directly i.e. this is not possible: 
```
model = model_class(config)
inputs = self._prepare_for_class(inputs_dict, model_class)
serving_outputs = model.serving(inputs)
```

Running this on vision models raises the following: 
```
E                           ValueError: The channel dimension of the inputs should be defined. The input_shape received is (None, None, None, None), where axis -1 (0-based) is the channel dimension, which found to be `None`.
```

This is because the input signature defined in the `tf.function` decorator for the `serving` method has all of the input dimensions defined as `None`:
```
 @tf.function(input_signature=[{
     "pixel_values": tf.TensorSpec((None, None, None, None), tf.float32, name="pixel_values"),
            }]
    )
    def serving(self, inputs):
        ....
```
We can't hard code the channel dimensions as e.g. `3` as we want to support both RGB and greyscale images (although testing this locally does work). 


**5. Moves `test_saved_model_creation` back into `test_modeling_tf_common` and add explicit skips**

There were quite a few models that couldn't be saved with `model.save_pretrained(path, saved_model=True)` and quite a few whose input signature or return types from `serving_output` were broken or inconsistent with the model. 

I think this is in part because the relevant test was moved to only be applied to certain core models, and those models didn't explicitly skip. See [#14415](https://github.com/huggingface/transformers/pull/14415), [#9478](https://github.com/huggingface/transformers/pull/9478)

I've:
* added it back to common so that it's added to models by default. CI currently running and passing, 
* added `unittest.skip` decorator so it's counted as a skipped rather than passed test on all models that were previously skipping.

**6. Update logic in models such that their graph can be created and saved.** 

Adds serving logic to enable saving of models and ensures their outputs are transformed in line with the rest of the library

## Fixes 
https://github.com/huggingface/transformers/issues/18179
https://github.com/huggingface/transformers/issues/18164
https://github.com/huggingface/transformers/issues/17233
https://github.com/huggingface/transformers/issues/17285
https://discuss.huggingface.co/t/tfresnetforimageclassification-fails-with-save-pretrained-when-saved-model-is-true/20404

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#start-contributing-pull-requests),
      Pull Request section?
- [ ] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [x] Did you write any new necessary tests?

amyeroberts committed:
Add serving_output and serving methods to some vision models

amyeroberts committed:
Add serving outputs for DeiT

HuggingFaceDocBuilderDev commented:
The docs for this PR live [here](https://moon-ci-docs.huggingface.co/docs/transformers/pr_18153). All of your documentation changes will be reflected on that endpoint.

amyeroberts committed:
Don't convert hidden states - differing shapes

amyeroberts committed:
Make saveable

amyeroberts committed:
Fix up

sayakpaul commented:
@amyeroberts thank you for digging into this one. An important one albeit. 

> We can't hard code the channel dimensions as e.g. 3 as we want to support both RGB and greyscale images (although testing this locally does work).

Could you shed more details on the locally testing you're referring to here?

amyeroberts committed:
Make swin saveable

amyeroberts committed:
Add in tests

amyeroberts committed:
Fix funnel tests (can't convert to tensor)

amyeroberts commented:
> Could you shed more details on the locally testing you're referring to here?

@sayakpaul Sure :) In the test `test_prepare_serving_output`, if serving outputs are got by calling `serving` directly i.e. 
```
            inputs = self._prepare_for_class(inputs_dict, model_class)
            serving_outputs = model.serving(inputs)
```
This will fail for vision models with the channel dim error I posted above. If I changed the input signature such that channel dimension is hard coded then it runs e.g. 

```
    @tf.function(
        input_signature=[
            {
                "pixel_values": tf.TensorSpec((None, 3, None, None), tf.float32, name="pixel_values"),
            }
        ]
    )
```

I run the test with 
```pytest tests/models/{model_name}/test_modeling_tf_{model_name}.py::TF{ModelName}ModelTest::test_prepare_serving_output```

sayakpaul commented:
Thanks. A follow-up question. So if you run with (`pytest tests/models/{model_name}/test_modeling_tf_{model_name}.py::TF{ModelName}ModelTest::test_prepare_serving_output`) how does it get the hardcoded value for channels? Or do you first hard-code it and then run it?

amyeroberts committed:
Fix numpy call

amyeroberts commented on src/transformers/models/hubert/modeling_tf_hubert.py:
```diff
@@ -227,7 +227,7 @@ def _compute_mask_indices(
         )
     # compute number of masked spans in batch
-    num_masked_spans = int(mask_prob * sequence_length / mask_length + tf.random.uniform((1,)))
+    num_masked_spans = int(mask_prob * sequence_length / mask_length + np.random.uniform(size=(1,)))
```
Using `tf.random.uniform` here creates issues with int casting here, `max` on L231 and other logic that assumes an int downstream.

sayakpaul replied:
@amyeroberts won't using `np` create problems in graph mode? For e.g. if the model was compiled?

amyeroberts replied:
Yep. I've removed it and added all the necessary tf logic.

amyeroberts commented on src/transformers/models/convnext/modeling_tf_convnext.py:
```diff
@@ -383,7 +383,8 @@ def serving(self, inputs):
             inputs (`Dict[str, tf.Tensor]`):
                 The input of the saved model as a dictionary of tensors.
         """
-        return self.call(inputs)
```
Adding in `self.serving(output)` return to: 
1. keep in line with the rest of the models call pattern and return types e.g. not including loss (I don't know why this is done) 
2. allow for testing (see PR desc.)

amyeroberts commented on src/transformers/models/funnel/modeling_tf_funnel.py:
```diff
@@ -1127,12 +1127,12 @@ def call(
             training=training,
         )
 
-    # Copied from transformers.models.distilbert.modeling_tf_distilbert.TFDistilBertModel.serving_output
```
Removing copied from across models as the tensors in the tuples for hidden_states and activations are different sizes so can't be called with `tf.convert_to_tensor`

ydshieh replied:
A quick look in Funnel docstring, I could not find why they have different sizes.

https://github.com/huggingface/transformers/blob/2c5747edfe383eee073119de784fa148befe9f2d/src/transformers/models/funnel/modeling_funnel.py#L843-L845

Maybe the docstring is wrong, and we should update it (in another PR for sure) ..?

amyeroberts replied:
It seems because it uses pooling, the sequence length is different after each block: https://github.com/huggingface/transformers/blob/2c5747edfe383eee073119de784fa148befe9f2d/docs/source/en/model_doc/funnel.mdx?plain=1#L38

amyeroberts committed:
Tidy up a bit

amyeroberts committed:
Add in hidden states - resnet

amyeroberts commented on src/transformers/models/speech_to_text/modeling_tf_speech_to_text.py:
```diff
@@ -143,7 +143,9 @@ def __init__(self, config: Speech2TextConfig, **kwargs):
 
     def call(self, input_features: tf.Tensor) -> tf.Tensor:
-        hidden_states = tf.identity(input_features)  # TF Conv1D assumes Batch x Time x Channels, same as the input
+        hidden_states = tf.cast(
+            input_features, tf.float32
+        )  # TF Conv1D assumes Batch x Time x Channels, same as the input
```
I'm not sure what the `tf.identity` was doing here - will removing this break anything @gante?

Casting to float to enable the `tf.concat` in graph mode in L156

gante replied:
If I remember correctly, this line only existed to rename the variable (and probably `hidden_states = input_features` would be enough for that 🤔 )

amyeroberts replied:
Could it be a mutation thing? tf.identity returns a copy

gante replied:
To be frank I can't remember if that was the cause :D But since it was my first HF PR, I would err towards a bad choice on my end

amyeroberts commented on src/transformers/models/wav2vec2/modeling_tf_wav2vec2.py:
```diff
@@ -268,7 +268,7 @@ def _compute_mask_indices(
         )
     # compute number of masked spans in batch
-    num_masked_spans = int(mask_prob * sequence_length / mask_length + tf.random.uniform((1,)))
+    num_masked_spans = int(mask_prob * sequence_length / mask_length + np.random.uniform(size=(1,)))
```
Same as with hubert - changing to numpy call to enable int cast in graph mode (and allow downstream computations that assume int).

amyeroberts commented on src/transformers/models/swin/modeling_tf_swin.py:
```diff
@@ -227,7 +227,7 @@ def window_reverse(windows: tf.Tensor, window_size: int, height: int, width: int
     """
     x = shape_list(windows)[0]
-    y = tf.cast(height * width / window_size / window_size, tf.int32)
+    y = tf.cast(height * width / (window_size * window_size), tf.int32)
```
The previous calculation fails because: 
* `height * width / widow_size` is calculated first, before the final `window_size` division. 
* `height`, `width` are ints, `window_size` is a tensor of type int
* Therefore, the resulting `height * width / widow_size` is a tensor of type float
* This fails when dividing by `window_size` again as the tensors are of different types

amyeroberts commented on src/transformers/models/swin/modeling_tf_swin.py:
```diff
@@ -1396,7 +1414,7 @@ def __init__(self, config: SwinConfig):
             tf.keras.layers.Dense(config.num_labels, name="classifier")
             if config.num_labels > 0
-            else tf.identity(name="classifier")
+            else tf.keras.layers.Activation("linear", name="classifier")
```
This was a bug

amyeroberts committed:
Remove numpy

amyeroberts committed:
Fix failing tests - tensor shape and skipping tests

amyeroberts commented on src/transformers/models/speech_to_text/modeling_tf_speech_to_text.py:
```diff
@@ -187,23 +189,22 @@ def _get_embedding(num_embeddings: int, embedding_dim: int, padding_idx: Optiona
+        self.embeddings = self.add_weight(
+            name="weights",  # name also used in PT
+            shape=shape_list(self.embedding_weights),
+            trainable=False,
```
As far as I can tell from PT model, this should be fixed. Is this right?

gante replied:
Right, it is a non-trainable positional embedding that is obtained from sines and cosines 👍

amyeroberts commented on src/transformers/models/speech_to_text/modeling_tf_speech_to_text.py:
```diff
@@ -942,22 +942,6 @@ def get_embed_tokens(self):
     def set_embed_tokens(self, embed_tokens):
         self.embed_tokens = embed_tokens
 
-    def _prepare_decoder_attention_mask(self, attention_mask, input_shape, inputs_embeds, past_key_values_length):
```
This function causes issues in graph mode, as `combined_attention_mask` can be either a tensor or None. Instead, the equivalent logic in bart call method is used in this model's to safely create `combined_attention_mask`

amyeroberts commented on src/transformers/models/swin/modeling_tf_swin.py:
```diff
@@ -659,64 +659,46 @@ def __init__(
             TFSwinDropPath(config.drop_path_rate, name="drop_path")
             if config.drop_path_rate > 0.0
-            else tf.identity(name="drop_path")
+            else tf.keras.layers.Activation("linear", name="drop_path")
```
This was a bug

amyeroberts commented on src/transformers/models/hubert/modeling_tf_hubert.py:
```diff
@@ -227,12 +227,17 @@ def _compute_mask_indices(
     # compute number of masked spans in batch
-    num_masked_spans = int(mask_prob * sequence_length / mask_length + tf.random.uniform((1,)))
-    num_masked_spans = max(num_masked_spans, min_masks)
+    num_masked_spans = mask_prob * sequence_length / mask_length + tf.random.uniform((1,))
```
we can't directly use `int` on the resulting `tf.Tensor` so we cast later

amyeroberts commented on src/transformers/models/hubert/modeling_tf_hubert.py:
```diff
@@ -227,12 +227,17 @@ def _compute_mask_indices(
     # make sure num masked indices <= sequence_length
-    if num_masked_spans * mask_length > sequence_length:
-        num_masked_spans = sequence_length // mask_length
+    num_masked_spans = tf.cond(
```
In graph mode, we can't used a tensor in an if/else statement. You get the following error:
`OperatorNotAllowedInGraphError: using a `tf.Tensor` as a Python `bool` is not allowed: AutoGraph did convert this function. This might indicate you are trying to use an unsupported feature.`

amyeroberts replied:
Good shout. Updated 👍

amyeroberts commented on src/transformers/models/swin/modeling_tf_swin.py:
```diff
@@ -659,64 +659,46 @@ def __init__(
         self.intermediate = TFSwinIntermediate(config, dim, name="intermediate")
         self.swin_output = TFSwinOutput(config, dim, name="output")
 
-    def set_shift_and_window_size(self, input_resolution: Tuple[int, int]) -> None:
```
Conditionally resetting the instance variables in both the `__init__` and `call` methods was not compatible with the model in graph mode. Instead `self.shift_size` and `self.window_size` are set in `__init__` with the same logic as this method.

In the `call` method, the local variables `shift_size` and `window_size` are set according to the same logic as this method. **Technically this isn't exactly equivalent** as the condition is dependant on `self.window_size` and in the original logic, `self.window_size` could change after every model call. As our model, at prediction time, should produce the same outputs if called twice on the same inputs, I think the refactor is consistent with expected behaviour.

amyeroberts commented on src/transformers/models/swin/modeling_tf_swin.py:
```diff
@@ -659,64 +659,46 @@ def __init__(
-                (-self.window_size, -self.shift_size),
-                (-self.shift_size, -1),
-            )
+    def get_attn_mask(self, height: int, width: int, window_size: int, shift_size: int) -> Optional[tf.Tensor]:
```
The previous `get_attn_mask` couldn't be used in graph mode, as the returned `attn_mask` could be a tensor or None. 

Similar to the reworking in `modeling_tf_speech_to_text`, an attention mask of all zeros, equivalent to having no mask, is created instead of `None` if `shift_size > 0`. 

Linked to deletion of `set_shift_and_window_size`, `shift_size` and `window_size` now passed as instance variables aren't reset at call time.

amyeroberts commented on tests/test_modeling_tf_common.py:
```diff
@@ -205,6 +205,46 @@ def test_save_load_config(self):
 
             self.assert_outputs_same(after_outputs, outputs)
 
+    def test_saved_model_creation(self):
```
I added this test back into `test_modeling_tf_common` as: 

* Most models when testing took <30 sec
* Many models couldn't be saved with `saved_model=True` flag and their serving code wouldn't have worked. I think it's better to include this as default and skip if too long. This increases the chance of catching issues and makes the test more visible.

amyeroberts committed:
Remove duplicated function

sgugger commented on src/transformers/models/data2vec/modeling_tf_data2vec_vision.py:
```diff
@@ -910,6 +910,17 @@ def call(
         return outputs
 
+    def serving_output(self, output: TFData2VecVisionModelOutputWithPooling) -> TFData2VecVisionModelOutputWithPooling:
+        hs = tf.convert_to_tensor(output.hidden_states) if self.config.output_hidden_states else None
```
Not a big fan of those two names, but even less fan of having the code span several lines. So if `hiddens` and `attentions`  still fit in one line, please use those (here and in the others) otherwise feel free to ignore.

amyeroberts replied:
Agreed - updated across the model files this PR touches 👍

sgugger reviewed (approved):
Thanks for diving into this! I'll let @gante and @Rocketknight1 dive deeper in the TensorFlow subtleties, but LGTM!

amyeroberts committed:
PR comments - formatting and var names

amyeroberts commented:
> Thanks. A follow-up question. So if you run with (`pytest tests/models/{model_name}/test_modeling_tf_{model_name}.py::TF{ModelName}ModelTest::test_prepare_serving_output`) how does it get the hardcoded value for channels? Or do you first hard-code it and then run it?

@sayakpaul Yes, I hardcode then run the tests. In this case, they pass.

amyeroberts commented on src/transformers/models/wav2vec2/modeling_tf_wav2vec2.py:
```diff
@@ -268,12 +268,17 @@ def _compute_mask_indices(
     # make sure num masked indices <= sequence_length
-    if num_masked_spans * mask_length > sequence_length:
-        num_masked_spans = sequence_length // mask_length
+    num_masked_spans = tf.cond(
```
Same as for hubert. In graph mode, we can't used a tensor in an if/else statement. You get the following error:
OperatorNotAllowedInGraphError: using a tf.Tensoras a Pythonbool is not allowed: AutoGraph did convert this function. This might indicate you are trying to use an unsupported feature.

gante replied:
(same comment as for hubert)

gante commented on src/transformers/models/speech_to_text/modeling_tf_speech_to_text.py:
```diff
@@ -187,23 +188,22 @@ def _get_embedding(num_embeddings: int, embedding_dim: int, padding_idx: Optiona
         if padding_idx is not None:
-            emb = tf.concat([emb[:padding_idx, :], tf.zeros((1, emb.shape[1])), emb[padding_idx + 1 :, :]], axis=0)
+            emb = tf.concat(
+                [emb[:padding_idx, :], tf.zeros((1, shape_list(emb)[1])), emb[padding_idx + 1 :, :]], axis=0
```
Same comment as above for `shape_list` -- `tf.shape(emb)` is a potential replacement

amyeroberts replied:
Updated :)

gante commented on src/transformers/models/speech_to_text/modeling_tf_speech_to_text.py:
```diff
@@ -187,23 +189,22 @@ def _get_embedding(num_embeddings: int, embedding_dim: int, padding_idx: Optiona
         Build shared token embedding layer Shared weights logic adapted from
         https://github.com/tensorflow/models/blob/a009f4fb9d2fc4949e32192a944688925ef78659/official/transformer/v2/embedding_layer.py#L24
         """
-        self._resize_embeddings()
```
Yup, embeddings in general have to be rewritten -- they still use TF 1 code :(

gante commented on src/transformers/models/speech_to_text/modeling_tf_speech_to_text.py:
```diff
@@ -779,7 +779,7 @@ def _get_feat_extract_output_lengths(self, input_lengths: tf.Tensor):
         # generate creates 3D attention mask, because of the shape of input_features
         # convert it to 2D if thats the case
-        if len(attention_mask.shape) > 2:
+        if len(shape_list(attention_mask)) > 2:
```
Same comment for `shape_list`. Here the issue is that `.shape` is not iterable in graph mode, but `.rank` should be okay.

amyeroberts replied:
Updated :)

gante commented on src/transformers/models/wav2vec2/modeling_tf_wav2vec2.py:
```diff
@@ -297,7 +302,7 @@ def _compute_mask_indices(
     # scatter indices to mask
     spec_aug_mask = _scatter_values_on_batch_indices(
-        tf.ones_like(spec_aug_mask_idxs), spec_aug_mask_idxs, spec_aug_mask.shape
+        tf.ones_like(spec_aug_mask_idxs), spec_aug_mask_idxs, shape_list(spec_aug_mask)
```
`tf.shape(spec_aug_mask)` if possible :)

amyeroberts replied:
Updated :)

gante commented on tests/models/bart/test_modeling_tf_bart.py:
```diff
@@ -293,8 +293,8 @@ def _get_word_embedding_weight(model, embedding_layer):
                                 models_equal = False
                     self.assertTrue(models_equal)
 
+    @unittest.skip(reason="Test takes too long to run (>30sec) causing the CI to fail.")
```
We also have a `@tooslow` decorator for this case (also applicable to other tests with the same skip reason)

https://github.com/huggingface/transformers/blob/ccc089780415445768bcfd3ac4418cec20353484/src/transformers/testing_utils.py#L219

amyeroberts replied:
Didn't know that - thanks for pointing out! I've updated the tests

gante reviewed (approved):
In general, looks good to me 👍 Thanks for fixing all these failing cases!

P.S.: the changes to swin were pretty big and probably should have been a PR on their own, but I'm also often guilty of this sin :D Have you double-checked the slow tests for the models with the most changes?

ydshieh commented:
Before merge, could you measure the timing for the tests `test_saved_model_creation` on **CPU**? You can run like

```python
python -m pytest -v tests -k "test_saved_model_creation" --durations=0 --make-reports=tests_timing
```
and copy-paste the results from `reports/tests_timing/durations.txt `

sgugger commented:
@ydshieh We don't run the slow tests on CPU, only on GPU/multi-GPU.

ydshieh commented:
> @ydshieh We don't run the slow tests on CPU, only on GPU/multi-GPU.

I should double check the latest version. My memory was in a previoius commit `Add in tests (505cb774b1b7eb5c9a6c8e2bc63f12061824b8bd)` while I asked the question 😢

ydshieh reviewed (commented):
Thank you @amyeroberts for this great work! I left a few questions.

ydshieh commented on src/transformers/models/hubert/modeling_tf_hubert.py:
```diff
@@ -1319,7 +1324,15 @@ def __init__(self, config, *inputs, **kwargs):
+    @tf.function(
+        input_signature=[
+            {
+                "input_values": tf.TensorSpec((None, None), tf.int32, name="input_ids"),
```
Also, the `name` should be changed too.

amyeroberts replied:
Updated the name 👍  Following your comment below, input_values are set to float and attention_maks and token_types_ids as int.

amyeroberts committed:
PR comments
Add suggestions made by Joao Gante:
* Use tf.shape instead of shape_list
* Use @tooslow decorator on tests
* Simplify some of the logic

amyeroberts committed:
PR comments
Address Yih-Dar Sheih comments - making tensor names consistent and make types float

amyeroberts commented on src/transformers/models/speech_to_text/modeling_tf_speech_to_text.py:
```diff
@@ -143,7 +143,9 @@ def __init__(self, config: Speech2TextConfig, **kwargs):
-        hidden_states = tf.identity(input_features)  # TF Conv1D assumes Batch x Time x Channels, same as the input
+        hidden_states = tf.cast(
+            input_features, tf.float32
+        )  # TF Conv1D assumes Batch x Time x Channels, same as the input
```
Done

amyeroberts commented on src/transformers/models/hubert/modeling_tf_hubert.py:
```diff
@@ -256,7 +261,7 @@ def _compute_mask_indices(
     # scatter indices to mask
     spec_aug_mask = _scatter_values_on_batch_indices(
-        tf.ones_like(spec_aug_mask_idxs), spec_aug_mask_idxs, spec_aug_mask.shape
+        tf.ones_like(spec_aug_mask_idxs), spec_aug_mask_idxs, shape_list(spec_aug_mask)
```
Updated :)

amyeroberts committed:
Types consistent with docs; disable test on swin (slow)

amyeroberts commented:
@ydshieh I know from @sgugger's comment, we don't run on CPU, but I ran the tests for reference (Macbook Pro 2021 M1 Max)
Based on this I disabled the test for Swin. The slowest tests  - `test_saved_model_creation_extended` - are independent of this PR. 

```
slowest durations
242.92s call     tests/models/convbert/test_modeling_tf_convbert.py::TFConvBertModelTest::test_saved_model_creation_extended
202.38s call     tests/models/bert/test_modeling_tf_bert.py::TFBertModelTest::test_saved_model_creation_extended
177.82s call     tests/models/gptj/test_modeling_tf_gptj.py::TFGPTJModelTest::test_saved_model_creation_extended
150.56s call     tests/models/bart/test_modeling_tf_bart.py::TFBartModelTest::test_saved_model_creation_extended
82.65s call     tests/models/gpt2/test_modeling_tf_gpt2.py::TFGPT2ModelTest::test_saved_model_creation_extended
61.12s call     tests/models/lxmert/test_modeling_tf_lxmert.py::TFLxmertModelTest::test_saved_model_creation_extended
26.88s call     tests/models/swin/test_modeling_tf_swin.py::TFSwinModelTest::test_saved_model_creation
20.14s call     tests/models/clip/test_modeling_tf_clip.py::TFCLIPTextModelTest::test_saved_model_creation_extended
19.51s call     tests/models/convbert/test_modeling_tf_convbert.py::TFConvBertModelTest::test_saved_model_creation
19.40s call     tests/models/gptj/test_modeling_tf_gptj.py::TFGPTJModelTest::test_saved_model_creation
18.54s call     tests/models/deberta_v2/test_modeling_tf_deberta_v2.py::TFDebertaModelTest::test_saved_model_creation
16.77s call     tests/models/speech_to_text/test_modeling_tf_speech_to_text.py::TFSpeech2TextModelTest::test_saved_model_creation
16.72s call     tests/models/clip/test_modeling_tf_clip.py::TFCLIPVisionModelTest::test_saved_model_creation_extended
14.51s call     tests/models/roformer/test_modeling_tf_roformer.py::TFRoFormerModelTest::test_saved_model_creation
14.37s call     tests/models/deberta/test_modeling_tf_deberta.py::TFDebertaModelTest::test_saved_model_creation
13.15s call     tests/models/wav2vec2/test_modeling_tf_wav2vec2.py::TFWav2Vec2ModelTest::test_saved_model_creation
13.12s call     tests/models/hubert/test_modeling_tf_hubert.py::TFHubertModelTest::test_saved_model_creation
12.79s call     tests/models/mpnet/test_modeling_tf_mpnet.py::TFMPNetModelTest::test_saved_model_creation
12.47s call     tests/models/tapas/test_modeling_tf_tapas.py::TFTapasModelTest::test_saved_model_creation
12.16s call     tests/models/layoutlm/test_modeling_tf_layoutlm.py::TFLayoutLMModelTest::test_saved_model_creation
12.16s call     tests/models/electra/test_modeling_tf_electra.py::TFElectraModelTest::test_saved_model_creation
12.12s call     tests/models/xlm/test_modeling_tf_xlm.py::TFXLMModelTest::test_saved_model_creation
11.92s call     tests/models/flaubert/test_modeling_tf_flaubert.py::TFFlaubertModelTest::test_saved_model_creation
11.54s call     tests/models/transfo_xl/test_modeling_tf_transfo_xl.py::TFTransfoXLModelTest::test_saved_model_creation
11.42s call     tests/models/bert/test_modeling_tf_bert.py::TFBertModelTest::test_saved_model_creation
11.36s call     tests/models/rembert/test_modeling_tf_rembert.py::TFRemBertModelTest::test_saved_model_creation
11.12s call     tests/models/distilbert/test_modeling_tf_distilbert.py::TFDistilBertModelTest::test_saved_model_creation
10.85s call     tests/models/ctrl/test_modeling_tf_ctrl.py::TFCTRLModelTest::test_saved_model_creation
10.77s call     tests/models/roberta/test_modeling_tf_roberta.py::TFRobertaModelTest::test_saved_model_creation
10.51s call     tests/models/wav2vec2/test_modeling_tf_wav2vec2.py::TFWav2Vec2RobustModelTest::test_saved_model_creation
10.44s call     tests/models/gpt2/test_modeling_tf_gpt2.py::TFGPT2ModelTest::test_saved_model_creation
10.30s call     tests/models/hubert/test_modeling_tf_hubert.py::TFHubertRobustModelTest::test_saved_model_creation
10.23s call     tests/models/deit/test_modeling_tf_deit.py::TFDeiTModelTest::test_saved_model_creation
10.20s call     tests/models/t5/test_modeling_tf_t5.py::TFT5EncoderOnlyModelTest::test_saved_model_creation
10.14s call     tests/models/albert/test_modeling_tf_albert.py::TFAlbertModelTest::test_saved_model_creation
10.14s call     tests/models/clip/test_modeling_tf_clip.py::TFCLIPTextModelTest::test_saved_model_creation
9.53s call     tests/models/openai/test_modeling_tf_openai.py::TFOpenAIGPTModelTest::test_saved_model_creation
9.11s call     tests/models/vit_mae/test_modeling_tf_vit_mae.py::TFViTMAEModelTest::test_saved_model_creation
8.71s call     tests/models/vit/test_modeling_tf_vit.py::TFViTModelTest::test_saved_model_creation
8.68s call     tests/models/clip/test_modeling_tf_clip.py::TFCLIPVisionModelTest::test_saved_model_creation
8.44s call     tests/models/dpr/test_modeling_tf_dpr.py::TFDPRModelTest::test_saved_model_creation
8.14s call     tests/models/xlnet/test_modeling_tf_xlnet.py::TFXLNetModelTest::test_saved_model_creation
7.54s call     tests/models/data2vec/test_modeling_tf_data2vec_vision.py::TFData2VecVisionModelTest::test_saved_model_creation
6.70s call     tests/models/convnext/test_modeling_tf_convnext.py::TFConvNextModelTest::test_saved_model_creation
6.21s call     tests/models/regnet/test_modeling_tf_regnet.py::TFRegNetModelTest::test_saved_model_creation
4.62s call     tests/models/resnet/test_modeling_tf_resnet.py::ResNetModelTest::test_saved_model_creation
```

amyeroberts commented on src/transformers/models/speech_to_text/modeling_tf_speech_to_text.py:
```diff
@@ -608,7 +608,7 @@ def _get_feat_extract_output_lengths(self, input_lengths: tf.Tensor):
         input_signature=[
             {
-                "input_ids": tf.TensorSpec((None, None), tf.int32, name="input_ids"),
+                "input_features": tf.TensorSpec((None, None, None), tf.int32, name="input_ids"),
```
I've updated the name. 
For the types, from the pytorch model, `input_features`,  `decoder_inputs` and `decoder_attention_mask` should be int32. `attention_mask` isn't specified, but `test_saved_model_creation` fails if it's set to float32.

ydshieh replied:
From the 2 places

https://github.com/huggingface/transformers/blob/2c5747edfe383eee073119de784fa148befe9f2d/tests/test_modeling_common.py#L1720
and 
https://github.com/huggingface/transformers/blob/2c5747edfe383eee073119de784fa148befe9f2d/tests/test_modeling_tf_common.py#L531
as well as the docstring (the 2nd line below)
https://github.com/huggingface/transformers/blob/2c5747edfe383eee073119de784fa148befe9f2d/src/transformers/models/speech_to_text/modeling_speech_to_text.py#L602
I still think it should be float. The part `torch.LongTensor` seems to be a mistake.

(cc @patrickvonplaten @anton for a confirmation.)

amyeroberts replied:
OK, that's a good point. I've updated it to float32 (passes tests locally)

amyeroberts commented on src/transformers/models/wav2vec2/modeling_tf_wav2vec2.py:
```diff
@@ -1352,7 +1357,15 @@ def __init__(self, config, *inputs, **kwargs):
+    @tf.function(
+        input_signature=[
+            {
+                "input_values": tf.TensorSpec((None, None), tf.int32, name="input_ids"),
```
Updated the name 👍 Following your comment below, input_values are set to float and attention_maks and token_types_ids as int.

amyeroberts committed:
CI trigger

amyeroberts committed:
Change input_features to float32

gante commented:
I noticed a cool thing -- assuming the models with `@tooslow` also pass the tests (which I'm assuming they do, from [this](https://github.com/huggingface/transformers/pull/18153#issuecomment-1191706099) comment), this PR fixes https://github.com/huggingface/transformers/issues/17233, as we know we can create a `SavedModel` (which contains a graph) for all models. The problematic line that made me open the issue was actually rewritten in this PR.

@amyeroberts can you add this issue to your `Fixes` list above? :D

amyeroberts commented:
@gante Nice spot :D Yep - I just double checked, and all the models with the @tooslow decorator can be saved with `saved_model=True` - and so their graph can be built. Added the fixes. 

The only model we can't save at the moment is CLIP, due to the nested dict of outputs.

amyeroberts committed:
Merge branch 'main' into fix-tf-resnet-saving-bug

amyeroberts committed:
Add serving_output for segformer

amyeroberts committed:
Fixup

amyeroberts commented:
@ydshieh @Rocketknight1 Am I OK to merge?

ydshieh reviewed (approved):
Thanks again!

Rocketknight1 reviewed (approved):
Overall, this is an extremely strong PR. The model changes were comprehensive and I'm impressed that you got the `SavedModel` test back into common!
